netlify-functions/src/main/scala/feral/netlify/IOFunction.scala

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright 2021 Typelevel
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package feral.netlify
+
+import cats.effect.IO
+import cats.effect.IOLocal
+import cats.effect.kernel.Resource
+import feral.IOSetup
+import io.circe.Decoder
+import io.circe.Encoder
+import io.circe.scalajs._
+
+import scala.scalajs.js
+import scala.scalajs.js.JSConverters._
+import scala.scalajs.js.|
+
+abstract class IOFunction[Event, Result](
+    implicit private[netlify] val decoder: Decoder[Event],
+    private[netlify] val encoder: Encoder[Result]
+) extends IOSetup {
+
+  final type Setup = (Event, Context[IO]) => IO[Option[Result]]
+
+  final override protected def setup: Resource[IO, Setup] =
+    handler.map { handler => (event, context) =>
+      for {
+        event <- IOLocal(event)
+        context <- IOLocal(context)
+        env = FunctionEnv.ioLambdaEnv(event, context)
+        result <- handler(env)
+      } yield result
+    }
+
+  def main(args: Array[String]): Unit = {
+    // Netlify functions require the entrypoint to be called `handler`
+    js.Dynamic.global.exports.updateDynamic("handler")(handlerFn)
+  }
+
+  private lazy val handlerFn
+      : js.Function2[js.Any, facade.Context, js.Promise[js.Any | Unit]] = {
+    (event: js.Any, context: facade.Context) =>
```

Comment:
```suggestion
  @JSExportTopLevel("handler")
  def exportedHandler(event: js.Any, context: facade.Context): js.Promise[js.Any | Unit]] =
```

Reply:
"Only static objects may export their members to the top level"

Reply:
Dang. Thanks for trying.